apps/shared/avifpng.c

```diff
@@ -144,21 +160,20 @@ static avifBool avifExtractExif(png_structp png, png_infop const info, avifImage
     const png_uint_32 numTextChunks = png_get_text(png, info, &text, NULL);
     for (png_uint_32 i = 0; i < numTextChunks; ++i, ++text) {
         if (!strcmp(text->key, "Raw profile type exif") || !strcmp(text->key, "Raw profile type APP1")) {
-            png_size_t textLength;
-            switch (text->compression) {
-#ifdef PNG_iTXt_SUPPORTED
-                case PNG_ITXT_COMPRESSION_NONE:
-                case PNG_ITXT_COMPRESSION_zTXt:
-                    textLength = text->itxt_length;
-                    break;
-#endif
-                case PNG_TEXT_COMPRESSION_NONE:
-                case PNG_TEXT_COMPRESSION_zTXt:
-                default:
-                    textLength = text->text_length;
-                    break;
-            }
-            return avifCopyRawProfile(text->text, textLength, &avif->exif);
+            return avifCopyRawProfile("Exif", text->text, avifGetPngTextLength(text), &avif->exif);
+        }
+    }
+    return AVIF_TRUE;
+}
+
+// Same for XMP.
+static avifBool avifExtractXMP(png_structp png, png_infop info, avifImage * avif)
+{
+    png_textp text = NULL;
+    const png_uint_32 numTextChunks = png_get_text(png, info, &text, NULL);
+    for (png_uint_32 i = 0; i < numTextChunks; ++i, ++text) {
+        if (!strcmp(text->key, "Raw profile type xmp")) {
```

Comment:
There are a couple of other possible locations for XMP data (the most other important one being "XML:com.adobe.xmp"), see (especially the comments):

https://exiftool.org/TagNames/PNG.html

Also interesting there is/was this "Raw profile type APP1" chunk in the wild that could be either Exif or XMP.

Finally, there is also a possible alternative ICC chunk, haven't checked if libavif supports that one id addition to iCCP, or perhaps libpng checks for it as well with their ICC set/get methods?

Reply:
> There are a couple of other possible locations for XMP data (the most other important one being "XML:com.adobe.xmp"

Done.

> the obsoleted tXMP

How can I read that with libpng? Can `png_get_text()` return a `png_text_struct::key` being "tXMP"? Do you have an example image? I did not find any `tXMP` among ~50k crawled PNG images (using `exiv2 -pS | grep "tXMP"`) so it is probably not worth the hassle.

> Also interesting there is/was this "Raw profile type APP1" chunk in the wild that could be either Exif or XMP.

Done, based on whether it begins with "II" or "MM". Everything else is mapped to XMP as there seems to be many possible XMP first bytes.

> Finally, there is also a possible alternative ICC chunk as text, haven't checked if libavif supports that one in addition to the dedicated iCCP, or perhaps libpng already checks for it as well with their ICC set/get methods?

Do you have an example image?

Reply:
> How can I read that with libpng? Can `png_get_text()` return a `png_text_struct::key` being "tXMP"? Do you have an example image? I did not find any `tXMP` among ~50k crawled PNG images (using `exiv2 -pS | grep "tXMP"`) so it is probably not worth the hassle.

[I guess only w/ the "unknown" chunk API](https://refspecs.linuxbase.org/LSB_4.1.0/LSB-Desktop-generic/LSB-Desktop-generic/libpng-png-get-unknown-chunks-1.html). But I agree, probably not worth the hassle if it really was pulled from the spec 20 years ago, and indeed, there are no examples around.

Same for the ICC alternative - as there are probably no test images in the exiv2 database and the iCCP chunk has been around for a while, probably not worth the hassle either until someone reports it.